### DISCUSSION TITLE
GRIB beginner: Trying to understand how to best work with layers

### DISCUSSION THREAD

**Discussion by shortcircuit3** [noritada/grib-rs#97] [Q&A]
Hello!

I'm new to working with GRIB files and I'm trying to use this library for extract necessary data form the file.

File: https://drive.google.com/file/d/1SNf2Wr450dkno8fy5rNJrnlfa4IHp9DR/view?usp=sharing

In the attached file, I'm looking to extract data from specific layers. In other libraries, I've used keys and abbreviations for data lookup, but I can' figure out how to properly target layers with this library

I have modified code from the examples directory that lists out the layers

```rs
use std::{
    any::Any, borrow::{Borrow, BorrowMut}, env, error::Error, fs::File, io::BufReader, path::Path
};

use grib::{codetables::{grib2::*, CodeTable4_2, Lookup}, ForecastTime, Name};

fn main() -> Result<(), Box<dyn Error>> {
    // Open the input file in a normal way.
    let f = File::open("src/gfswave.t18z.atlocn.0p16.f064.grib2")?;
    let f = BufReader::new(f);

    // Read with the reader.
    let grib2 = grib::from_reader(f)?;

    // Iterate over layers.
    for (_index, submessage) in grib2.iter() {
        let discipline = submessage.indicator().discipline;
        let category = submessage.prod_def().parameter_category().unwrap();
        let parameter_num = submessage.prod_def().parameter_number().unwrap();
        let parameter = grib::codetables::CodeTable4_2::new(discipline, category)
            .lookup(usize::from(parameter_num));

        // Only looking for param 8 for now
        if parameter_num == 8 {

            let latlons = submessage.latlons()?;

            let decoder = grib::Grib2SubmessageDecoder::from(submessage)?;

            let values = decoder.dispatch()?;

            for ((lat, lon), value) in latlons.zip(values) {
                if !value.is_nan() {
                    println!("{lat} {lon} {value}");
                    println!("================");
                }
            }
        }

    }

    Ok(())
}
```

I see here that it's necessary to use code tables for lookups using `discipline`, `category`, and `parameter`.

However I can't seem to figure out how to properly filter using the following code.

```rs
    for (index, submessage) in grib2.iter() {
        let discipline = submessage.indicator().discipline;
        let category = submessage.prod_def().parameter_category().unwrap();
        let parameter_num = submessage.prod_def().parameter_number().unwrap();
        
        let a = grib::codetables::CodeTable4_2::new(discipline, category).lookup(usize::from(parameter_num));

        if discipline == 10 && category == 0 && parameter_num == 8 {
            let (first, _second) = submessage.prod_def().fixed_surfaces().unwrap();

            let elevation_level = first.value();
            
            println!("{}.{}\n{}\n{:?}", index.0, index.1, submessage.describe(), elevation_level);
        }

     
        // let ft = submessage.prod_def().forecast_time();
        // if matches!(
        //     ft,
        //     Some(ForecastTime {
        //         unit: Name(Table4_4::Hour),
        //         value: hours,
        //     }) if hours == forecast_time_hours
        // ) {
        //     println!("{}.{}\n{}", index.0, index.1, submessage.describe());
        // }
    }
```

Can someone help me understand how to use the api in a more sane way?

**Comment by noritada** [comment]
Thank you for raising the issue.
The weather element representation of submessages is one of the issues I am aware of.

About your quesition, if you want to filter submessages by weather element, we currently have two options.
as I can find the code fragment in your source code, 

- Determining if the code numbers (including the discipline and category numbers) of the element the submessage represents matches the code numbers of the weather element you want to use.
- Similarly, after obtaining the code numbers, converting them to a text representation and determine if it matches the text of the weather element you want to use.

The first one is what you show and the second is like this:

```rust
    for (index, submessage) in grib2.iter() {
        let discipline = submessage.indicator().discipline;
        let category = submessage.prod_def().parameter_category().unwrap();
        let parameter_num = submessage.prod_def().parameter_number().unwrap();

        let description = grib::codetables::CodeTable4_2::new(discipline, category)
            .lookup(usize::from(parameter_num))
            .to_string();

        if description == "Significant height of swell waves" {
             todo!()
        }
    }
```

The first method requires knowledge of GRIB2 code tables.
The second method is a text match and does not require knowledge of code tables, but has the disadvantage that the element to be searched for cannot be filtered without knowing how that element is represented as text. Another problem is that the code to obtain the textual representation is currently a bit complex.

I am exploring what kind of API to provide to make it easier to use while implementing other features.
At the moment I have two ideas.

- A. Allow users to use the “second way” described above more easily.
  For example, would you be happy if you could write `let Some(description) = submessage.description();` without having to extract `discipline`, `category`, and `parameter_num` to obtain description?

- B. Implement wgrib2's element codes.
  In wgrib2, each weather element is given its own short code, like [`SWELL`](https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table4-2-10-0.shtml).
  If we implement this code in this library as well, we can filter with the same code as wgrib2.

Thank you for the feedback!

**Reply by shortcircuit3** [reply]
Thank you so much for the answer! After researching other libraries as well as this one, my impression is that this library closely maps the way grib files actually work. Other libraries seem to have abstract functions to more easily fetch necessary data.

For example, the [pygrib api](https://jswhit.github.io/pygrib/api.html#example-usage) is quite simple and maps to my mental model for how to access data.

```py
grbs = pygrib.open('sampledata/flux.grb')
swell = grbs.select(shortName='SWELL', level=1)
swells = grbs.select(shortName='SWELL', level=(1,2,3))

# Read the first GRIB message
grib_message = grbs.message(1)

# Check if the GRIB message has a specific key
key_to_check = 'temperature'
if grib_message.haskey(key_to_check):
    print(f"The GRIB message has the key: {key_to_check}")
else:
    print(f"The GRIB message does not have the key: {key_to_check}")

```

I would love if there was a simple api to target different layers and parameters based on layer data.

However, your explanation answers my question that I need to loop over the layer to find what i'm looking for.

**Reply by noritada** [reply]
Sorry for the late reply. Thank you for your opinion.

I have been using pygrib for many years and strongly agree with your opinion, and one of our goals at grib-rs is to make it look that way in the future. On the other hand, it is true that we are not currently able to provide such an API because the amount of work would be enormous and we would need to consider it from the API design.

In GRIB2, althohgh the code table mechanism defines “descriptions” of values and associates numbers with meanings, there is no “concise string code system that is easy for humans to understand,” and tools such as wgrib2 and eccodes each define their own such string code system. Therefore, when creating a tool or library, it is necessary to define a string code that is easy for humans to understand. Right now, I am working on grib-rs to also allow elements to be specified with string codes defined for wgrib2 and eccodes, and since Rust is characterized by static typing, I am working on an implementation that allows string codes to be expressed as type names, which IDE can complete, rather than strings.

Another difficulty with GRIB2 is the template mechanism that allows data providers to vary the set of fields that can be included in the data, depending on the type of data. The mechanism you write about “being able to check for the presence of a field key” is necessary because the fields that can be included in the data are different for each type of data. For templates, as with code tables, there is only a “description” of each field in the fieldset, and thus the library needs to design the field names. In time, I hope to make the various metadata available to users, but the work will likely go beyond the code table.

**Reply by noritada** [reply]
In the near future, the crate will allow users to filter submessages like this:

```rust
grib2.iter().find(|submessage| {
    submessage.parameter().is_some_and(|param| param.is_identical_to(NCEP::HGT))
});
```

(See #100.)
